dipy/reconst/ivim.py

```diff
@@ -0,0 +1,391 @@
+""" Classes and functions for fitting ivim model """
+from __future__ import division, print_function, absolute_import
+import numpy as np
+
+from dipy.core.gradients import gradient_table
+from dipy.reconst.base import ReconstModel
+from dipy.reconst.dti import apparent_diffusion_coef
+from dipy.reconst.dti import TensorModel, mean_diffusivity
+from dipy.reconst.multi_voxel import multi_voxel_fit
+from dipy.reconst.vec_val_sum import vec_val_vect
+from dipy.core.sphere import Sphere
+
+from distutils.version import LooseVersion
+import scipy
+
+SCIPY_LESS_0_17 = LooseVersion(scipy.version.short_version) < '0.17'
+
+if not SCIPY_LESS_0_17:
+    least_squares = scipy.optimize.least_squares
+else:
+    leastsq = scipy.optimize.leastsq
+
+
+def ivim_function(params, bvals):
+    """The Intravoxel incoherent motion (IVIM) model function.
+
+    The IVIM model assumes that biological tissue includes a volume fraction
+    'f' of water flowing in perfused capillaries, with a perfusion
+    coefficient D* and a fraction (1-f) of static (diffusion only), intra and
+    extracellular water, with a diffusion coefficient D. In this model the
+    echo attenuation of a signal in a single voxel can be written as
+
+        .. math::
+
+        S(b) = S_0[f*e^{(-b*D\*)} + (1-f)e^{(-b*D)}]
+
+        Where:
+        .. math::
+
+        S_0, f, D* and D are the IVIM parameters.
+
+    Parameters
+    ----------
+    params : array
+        An array of IVIM parameters - S0, f, D_star, D
+
+    bvals : array
+        bvalues
+
+    References
+    ----------
+    .. [1] Le Bihan, Denis, et al. "Separation of diffusion
+               and perfusion in intravoxel incoherent motion MR
+               imaging." Radiology 168.2 (1988): 497-505.
+    .. [2] Federau, Christian, et al. "Quantitative measurement
+               of brain perfusion with intravoxel incoherent motion
+               MR imaging." Radiology 265.3 (2012): 874-881.
+    """
+    S0, f, D_star, D = params
+    S = S0 * (f * np.exp(-bvals * D_star) + (1 - f) * np.exp(-bvals * D))
+    return S
+
+
+def _ivim_error(params, bvals, signal):
+    """Error function to be used in fitting the IVIM model
+
+    Parameters
+    ----------
+    params : array
+        An array of IVIM parameters. [S0, f, D_star, D]
+
+    bvals : array
+        bvalues
+
+    signal : array
+        Array containing the actual signal values.
+
+    """
+    return (signal - ivim_function(params, bvals))
+
+
+class IvimModel(ReconstModel):
+    """Ivim model
+    """
+
+    def __init__(self, gtab, split_b=200.0,
+                 bounds=None, tol=1e-7,
+                 options={'gtol': 1e-7, 'ftol': 1e-7,
+                          'eps': 1e-7, 'maxiter': 1000}):
+        """
+        Initialize an IVIM model.
```

Comment:
As @arokem suggested before please remove the extensive explanation in the beginning of the ivim_function.